cdap-master/src/main/java/co/cask/cdap/data/tools/LogUpgrade.java

```diff
@@ -0,0 +1,148 @@
+          } catch (Exception e) {
+            throw Throwables.propagate(e);
+          }
+        }
+      });
+  }
+
+  @Override
+  public void upgrade(Injector injector) throws Exception {
+    final FileMetaDataManager fileMetaDataManager = new FileMetaDataManager(new LogSaverTableUtil(
+      nonNamespaedFramework, cConf), txClient, locationFactory);
+    logMDS.execute(new TransactionExecutor.Function<DatasetContext<OrderedTable>, Void>() {
+      @Override
+      public Void apply(DatasetContext<OrderedTable> ctx) throws Exception {
+        Scanner rows = ctx.get().scan(null, null);
+        Row row;
+        while ((row = rows.next()) != null) {
+          String key = Bytes.toString(Arrays.copyOfRange(row.getRow(), FileMetaDataManager.ROW_KEY_PREFIX.length,
+                                                         row.getRow().length)).trim();
+          Map<byte[], byte[]> allColumns = row.getColumns();
+          for (byte[] startTime : allColumns.keySet()) {
+            long startTimeMs = Bytes.toLong(startTime);
+            String oldPath = Bytes.toString(allColumns.get(startTime));
+            String newPath;
+            String newKey;
+            if (key.startsWith(Constants.Logging.SYSTEM_NAME) || key.startsWith(DEVELOPER_STRING)) {
+              newPath = upgradePath(key, oldPath);
+              newKey = upgradeKey(key);
+              renameLocation(new URI(oldPath).resolve("."), new URI(newPath).resolve("."));
+              fileMetaDataManager.writeMetaData(newKey, startTimeMs, locationFactory.create(new URI(newPath)));
+              LOG.info("Upgrading logs for {}", key);
+            } else if (!checkKeyValidality(key, OTHERS)) {
+              LOG.warn("Unknown key {} found. Failed to Upgrade", key);
+              throw new RuntimeException("Unknown key: " + key);
+            }
+          }
+        }
+        return null;
+      }
+    });
+  }
+
+  /**
+   * Adds the namespace in the key which is the log partition
+   *
+   * @param key the key which is the log partition
+   * @return the new key with namespace
+   */
+  private String upgradeKey(String key) {
+    if (key.startsWith(Constants.Logging.SYSTEM_NAME)) {
+      return key.replace(Constants.Logging.SYSTEM_NAME, Constants.SYSTEM_NAMESPACE);
+    }
+    return key.replace(DEVELOPER_STRING, Constants.DEFAULT_NAMESPACE);
+  }
+
+  /**
+   * Upgrades the path by adding the namespace in the log archive location
+   *
+   * @param key      the key which is log partition
+   * @param location the log archive location
+   * @return the new location which has the appropriate namespace
+   */
+  public String upgradePath(String key, String location) {
+    String newLocation = null;
+    if (key.startsWith(Constants.Logging.SYSTEM_NAME)) {
+      if (location != null) {
+        newLocation = location.substring(0, location.lastIndexOf(CDAP_WITH_FORWARD_SLASH)) +
```

Comment:
this is confusing to read, can you add a comment with an example of what this does